packages/patrol_cli/lib/src/features/drive/test_runner.dart

```diff
@@ -39,24 +65,52 @@ class TestRunner {
     if (_devices.isEmpty) {
       throw StateError('no devices to run tests on');
     }
-    if (_tests.isEmpty) {
-      throw StateError('no tests to run');
+    if (_targets.isEmpty) {
+      throw StateError('no test targets to run');
+    }
+
+    final builder = _builder;
+    if (builder == null) {
+      throw StateError('builder is null');
+    }
+
+    final executor = _executor;
+    if (executor == null) {
+      throw StateError('executor is null');
     }
 
     _running = true;
 
     final testRunsOnAllDevices = <Future<void>>[];
     for (final device in _devices.values) {
       Future<void> runTestsOnDevice() async {
-        for (final test in _tests) {
-          await test(device);
+        for (final target in _targets) {
+          if (_disposed) {
+            continue;
+          }
+
+          await builder(target, device);
+
+          for (var i = 0; i < _repeats; i++) {
+            if (_disposed) {
+              continue;
+            }
+
+            await executor(target, device);
+          }
```

Comment:
I wish there was a way to preempt a `Future`...

Reply:
What do you mean?

Reply:
I'd like to be able to cancel *any* work once `dispose()` is called'.

Reply:
I mean exactly what the creator of this issue meant: https://github.com/dart-lang/sdk/issues/42855